helix-term/src/handlers/document_colors.rs

```diff
@@ -0,0 +1,267 @@
+use std::{collections::HashSet, time::Duration};
+
+use helix_core::{syntax::LanguageServerFeature, text_annotations::InlineAnnotation};
+use helix_event::{cancelable_future, register_hook, TaskController};
+use helix_lsp::{
+    lsp::{self, ColorInformation},
+    OffsetEncoding,
+};
+use helix_view::{
+    document::DocumentColorSwatches,
+    events::{DocumentDidChange, DocumentDidOpen, LanguageServerExited, LanguageServerInitialized},
+    handlers::{lsp::DocumentColorsEvent, Handlers},
+    DocumentId, Editor, Theme, ViewId,
+};
+use tokio::time::Instant;
+
+use crate::job;
+
+#[derive(Default)]
+pub(super) struct DocumentColorsHandler {
+    views: HashSet<ViewId>,
+    docs: HashSet<DocumentId>,
+}
+
+impl DocumentColorsHandler {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+const DOCUMENT_CHANGE_DEBOUNCE: Duration = Duration::from_millis(50);
+
+impl helix_event::AsyncHook for DocumentColorsHandler {
+    type Event = DocumentColorsEvent;
+
+    fn handle_event(&mut self, event: Self::Event, _timeout: Option<Instant>) -> Option<Instant> {
+        let DocumentColorsEvent(doc_id) = event;
+        self.docs.insert(doc_id);
+        Some(Instant::now() + DOCUMENT_CHANGE_DEBOUNCE)
+    }
+
+    fn finish_debounce(&mut self) {
+        let mut views = std::mem::take(&mut self.views);
+        let docs = std::mem::take(&mut self.docs);
+
+        job::dispatch_blocking(move |editor, _compositor| {
+            editor
+                .handlers
+                .document_colors
+                .active_requests
+                .retain(|_, controller| controller.is_running());
+
+            // Drop any views which have been closed.
+            views.retain(|&view| editor.tree.contains(view));
+            // Add any views that show documents which changed.
+            views.extend(
+                editor
+                    .tree
+                    .views()
+                    .filter_map(|(view, _)| docs.contains(&view.doc).then_some(view.id)),
+            );
+
+            for view in views {
+                let doc = editor.tree.get(view).doc;
+                request_document_colors_for_view(editor, view, doc);
+            }
+        });
+    }
+}
+
+fn request_document_colors_for_view(editor: &mut Editor, view_id: ViewId, doc_id: DocumentId) {
+    if !editor.config().lsp.display_color_swatches {
+        return;
+    }
+
+    // Cancel any ongoing events for this view.
+    if let Some(mut controller) = editor
+        .handlers
+        .document_colors
+        .active_requests
+        .remove(&view_id)
+    {
+        controller.cancel();
+    }
+
+    let Some(doc) = editor.document(doc_id) else {
+        return;
+    };
+
+    let Some(language_server) = doc
+        .language_servers_with_feature(LanguageServerFeature::DocumentColors)
+        .next()
+    else {
+        return;
+    };
```

Comment:
This will only request colors from the first language server that can support the request. Instead we should use `doc.language_servers_with_feature` to create a `FuturesOrdered` - see uses of that type in the LSP commands module